File: packages/gis-common/src/gis_common/crs.py
# +-------------------------------------------------------------------------
#
#   地理智能平台 - 坐标参考与投影工具
#
#   文件:       crs.py
#
#   日期:       2026年04月14日
#   作者:       JamesLinYJ
# --------------------------------------------------------------------------

from __future__ import annotations

from functools import partial
from typing import Any

from pyproj import CRS, Transformer
from shapely.geometry import mapping, shape
from shapely.ops import transform


# CRS 与几何转换工具
#
# 提供局部米制投影选择、几何重投影以及 FeatureCollection 级别的坐标转换。
def shape_from_feature(feature: dict[str, Any]):
    return shape(feature["geometry"])


def feature_from_shape(geometry, properties: dict[str, Any] | None = None) -> dict[str, Any]:
    return {
        "type": "Feature",
        "properties": properties or {},
        "geometry": mapping(geometry),
    }


def choose_local_metric_epsg(longitude: float, latitude: float) -> int:
    zone = int((longitude + 180) / 6) + 1
    north = latitude >= 0
    return (32600 if north else 32700) + zone


def transform_geometry(geometry, src_epsg: int, dst_epsg: int):
    if src_epsg == dst_epsg:
        return geometry
    transformer = Transformer.from_crs(CRS.from_epsg(src_epsg), CRS.from_epsg(dst_epsg), always_xy=True)
    return transform(partial(transformer.transform), geometry)


def transform_feature_collection(payload: dict[str, Any], src_epsg: int, dst_epsg: int) -> dict[str, Any]:
    transformed_features = []
    for feature in payload.get("features", []):
        geom = shape_from_feature(feature)
        transformed_geom = transform_geometry(geom, src_epsg, dst_epsg)
        transformed_features.append(feature_from_shape(transformed_geom, dict(feature.get("properties", {}))))
    return {"type": "FeatureCollection", "features": transformed_features}

File: server/src/gis/crs.ts
// +-------------------------------------------------------------------------
//
//   地理智能平台 - CRS 与投影工具
//
//   文件:       crs.ts
//
//   日期:       2026年06月05日
//   作者:       JamesLinYJ
// --------------------------------------------------------------------------
import proj4 from 'proj4';
import type { Geometry } from './geojson.js';
// UTM 局部米制 EPSG
export function chooseLocalMetricEpsg(longitude: number, latitude: number): number {
    const zone = Math.floor((longitude + 180) / 6) + 1;
    const north = latitude >= 0;
    return (north ? 32600 : 32700) + zone;
}
// 重投影单个几何
export function transformGeometry<T extends Geometry>(geometry: T, srcEpsg: number, dstEpsg: number): T {
    if (srcEpsg === dstEpsg)
        return geometry;
    const fromProj = `EPSG:${srcEpsg}`;
    const toProj = `EPSG:${dstEpsg}`;
    function convert(coords: unknown): unknown {
        if (typeof coords === 'number')
            return coords;
        if (!Array.isArray(coords))
            return coords;
        if (coords.length === 2 && typeof coords[0] === 'number' && typeof coords[1] === 'number') {
            return proj4(fromProj, toProj, [coords[0], coords[1]]);
        }
        return coords.map(convert);
    }
    if (!('coordinates' in geometry)) return geometry;
    return { ...geometry, coordinates: convert(geometry.coordinates) } as T;
}
